### PR TITLE
Fix: 18183 spinner persists to magic code section

### PR DESCRIPTION
**Pull request by tienifr**
### Fixed Issues

$ https://github.com/Expensify/App/issues/18183
PROPOSAL: https://github.com/Expensify/App/issues/18183#issuecomment-1539371121

### Tests

1. Navigate to sing in page
2. Enter email and click on continue
3. Verify the spinner shouldn't persist to magic code section for the action taken on email section
4. Enter magic code
5. If 2FA is enable, verify the spinner shouldn't persist to 2fa section

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as above
### QA Steps

1. Navigate to sing in page
2. Enter email and click on continue
3. Verify the spinner shouldn't persist to magic code section for the action taken on email section
4. Enter magic code
5. If 2FA is enable, verify the spinner shouldn't persist to 2fa section

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/113963320/928b4151-8e52-4b1a-97fd-3c37946b8761

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/113963320/4521b73d-f316-4405-9d35-a3202faef84e

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/113963320/40ce9d99-97f5-435d-a65c-8b1303492b5e

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/113963320/2a611cf4-0d82-4156-bbec-dd11e4a98dbd

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/113963320/640076c3-81dc-4d42-97d8-d115aee0e161

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/113963320/7a1a2ec2-108f-4362-87ca-87138fc4cc6a

</details>

### CODE REVIEW

File: src/libs/SessionUtils.js [hunk 1]
```diff
@@ -28,7 +29,27 @@ function isLoggingInAsNewUser(transitionURL, sessionEmail) {
...
 }
 
+/**
+ * Indicates whether the button should show the loading indicator
+ * @param {String} loadingScreen
+ * @param {Boolean} requiresTwoFactorAuth
+ * @param {Boolean} isLoading
+ * @returns {Boolean}
+ */
+function shouldShowLoadingIndicator(loadingScreen, requiresTwoFactorAuth, isLoading) {
```
Comment (Julesssss) [inline src/libs/SessionUtils.js:39]: the `loadingScreen` param seems a bit out of place here, but I don't have any good suggestions for improvements. Maybe including a default or renaming `shouldShowLoadingIndicatorForScreen` could help make it clearer?

**Comment by tienifr** [comment]
> isLoading={this.props.account.isLoading && this.props.account.loadingForm === CONST.FORMS.VALIDATE_CODE_FORM}

@aimane-chnaif Thanks for your suggestion, but I wonder why the spinner doesn't persist to 2fa section once user enters magic code. isLoading and loadingForm will reset to null in successData/failureData at the same time

**Comment by tienifr** [comment]
When we click `Didn't receive a magic code?` isLoading in BaseValidateCodeForm or PasswordForm should be true right?

**Comment by tienifr** [comment]
changing loadingScreen to loadingForm is good
I also want to update the CONST like your suggestion

```
FORMS: {
        LOGIN_FORM: 'LoginForm',
        VALIDATE_CODE_FORM: 'ValidateCodeForm',
        RESEND_VALIDATION_FORM: 'ResendValidationForm',
        UNLINK_LOGIN_FORM: 'UnlinkLoginForm',
    },
```

But if we do: 

> isLoading = {this.props.account.isLoading && this.props.account.loadingForm === CONST.FORMS.VALIDATE_CODE_FORM}

When resend validation code, the loading indicator isn't shown. Is it OK?

https://github.com/Expensify/App/assets/113963320/7d38162d-37ee-47f6-9743-539930cdd968

Another problem:

1. Users enter validate code => isLoading = true, loadingForm = ValidateCodeForm
2. the loading indicator is shown
3. Be update requiresTwoFactorAuth = true
4. 2FA input is displayed while loading indicator persists // This is the problem
5. Reset isLoading = false and loadingForm = null in successData

That why I want to create a new function `shouldShowLoadingIndicatorForScreen` in my PR.

I want to know your thought @aimane-chnaif. Thanks

**Comment by tienifr** [comment]
@aimane-chnaif Thanks for your suggestion. I've updated my PR. I know the reason why you want to preserve isLoading in optimisticData, because it can be used in other places. I also tested all cases and there's no regressions came from my PR, but please help check and re-test.

**Review by aimane-chnaif — APPROVED** [review]
@Julesssss LGTM!

**Review by Julesssss — APPROVED** [review]
Nice work!

**Pull request merged**